### PR TITLE
Updates validation for dynamic data

### LINKED ISSUE #402: setting tags does not work after dynamic API
Azure requires that the following flow works:

GET resource
\<modify tags\>
**PUT** resource

At the moment I can see at least two issues:

1) non-delta validation on PUT/PATCH rejects the worker profiles (there's >1 of them, then we'll have issues at least with the name and count).  

I think we should substantially or maybe even entirely relax incoming non-delta validation of dynamic fields **after creation**.  Also need to consider the cluster version and client ID in this category.

2) delta validation on PUT/PATCH fails because it is comparing a non-enriched model with an incoming enriched model.  Delta validation on PUT/PATCH needs to be strong and continue to have strong test case coverage.  We should reject any deltas to dynamic fields.

We need to enrich the database model on PUT/PATCH before doing delta validation.

### PR DESCRIPTION
**Pull request by m1kola**
Fixes #402

### CODE REVIEW

File: pkg/frontend/openshiftcluster_putorpatch.go [hunk 1]
```diff
@@ -133,6 +134,10 @@ func (f *frontend) _putOrPatchOpenShiftCluster(ctx context.Context, r *http.Requ
 	if isCreate {
 		err = staticValidator.Static(ext, nil)
 	} else {
+		timeoutCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
+		defer cancel()
+		f.ocEnricher.Enrich(timeoutCtx, doc.OpenShiftCluster)
```
Comment (jim-minter) [inline pkg/frontend/openshiftcluster_putorpatch.go:139]: does this need to be pulled forward for the PATCH case?  i.e. we need to merge onto an enriched document?  To around line 105?

File: pkg/api/v20191231preview/openshiftcluster_validatestatic.go [hunk 2]
```diff
@@ -80,27 +80,29 @@ func (sv *openShiftClusterStaticValidator) validateProperties(path string, p *Op
 	default:
 		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, path+".provisioningState", "The provided provisioning state '%s' is invalid.", p.ProvisioningState)
 	}
+	if isCreate {
```
Comment (jim-minter) [inline pkg/api/v20191231preview/openshiftcluster_validatestatic.go:83]: * unit test coverage needed
* also want to see stronger unit tests for delta validation, e.g. cases where the length of WorkerProfiles changes

Comment (jim-minter) [inline pkg/api/v20191231preview/openshiftcluster_validatestatic.go:83]: currently the function flow very closely matches the field order in the struct; I think it is worth keeping that for maintainability.  I suggest multiple `if isCreate {}`s.

File: pkg/frontend/openshiftcluster_putorpatch.go (hunk 1 above)
Comment (jim-minter) [reply pkg/frontend/openshiftcluster_putorpatch.go:139]: This PR brings forward the caching requirement.
With this PR, if I do
PUT(create)->GET->GET->DELETE, at deleting I'll see the schema as I created it
if I do
PUT(create)->GET->PUT->GET->DELETE, at deleting I'll see the schema as I last PUT it

File: pkg/api/v20191231preview/openshiftcluster_validatestatic.go (hunk 2 above)
Comment (m1kola) [reply pkg/api/v20191231preview/openshiftcluster_validatestatic.go:83]: It is still a draft ;)

File: pkg/frontend/openshiftcluster_putorpatch.go (hunk 1 above)
Comment (jim-minter) [reply pkg/frontend/openshiftcluster_putorpatch.go:139]: TL;DR in a follow-up PR we should consider doing a DB write after all the enrich tasks have completed

File: pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go [hunk 3]
```diff
@@ -107,12 +108,12 @@ func runTests(t *testing.T, tests []*validateTest, delta bool) {
 				tt.modify(oc)
 			}
 
-			current := &api.OpenShiftCluster{}
+			var current *api.OpenShiftCluster
 			if delta {
+				current = &api.OpenShiftCluster{}
 				(&openShiftClusterConverter{}).ToInternal(validOpenShiftCluster(), current)
-			} else {
-				(&openShiftClusterConverter{}).ToInternal(oc, current)
```
Comment (m1kola) [inline pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go:114]: Previously in both `delta == true` and `delta == false` cases we were testing validator as if we were running an update. Now `delta == false` - is a create (`current == nil`) and `delta == true` is an update.

File: pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go [hunk 4]
```diff
@@ -699,6 +700,11 @@ func TestOpenShiftClusterStaticValidateDelta(t *testing.T) {
 			},
 			wantErr: "400: PropertyChangeNotAllowed: properties.masterProfile.subnetId: Changing property 'properties.masterProfile.subnetId' is not allowed.",
 		},
+		{
+			name:    "worker name change",
+			modify:  func(oc *OpenShiftCluster) { oc.Properties.WorkerProfiles[0].Name = "new-name" },
+			wantErr: "400: PropertyChangeNotAllowed: properties.workerProfiles['new-name'].name: Changing property 'properties.workerProfiles['new-name'].name' is not allowed.",
```
Comment (m1kola) [inline pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go:706]: Why does `immutable.Validate()` produces errors where istead of index it prints a value from the `Name` field?

https://github.com/Azure/ARO-RP/blob/389628e3b697d34a449e2b795258c5cdca43a169/pkg/util/immutable/immutable.go#L86-L91

File: pkg/api/v20191231preview/openshiftcluster_validatestatic.go [hunk 5]
```diff
@@ -131,8 +135,7 @@ func (sv *openShiftClusterStaticValidator) validateClusterProfile(path string, c
 		strings.ContainsRune(strings.TrimSuffix(cp.Domain, "."+sv.domain), '.') {
 		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, path+".domain", "The provided domain '%s' is invalid.", cp.Domain)
 	}
-	if isCreate && cp.Version != version.OpenShiftVersion ||
-		!isCreate && !validate.RxOpenShiftVersion.MatchString(cp.Version) {
```
Comment (m1kola) [inline pkg/api/v20191231preview/openshiftcluster_validatestatic.go:135]: I think, we do not need `validate.RxOpenShiftVersion` anymore. Am I wrong?

File: pkg/api/v20191231preview/openshiftcluster_validatestatic.go [hunk 6]
```diff
@@ -86,20 +86,24 @@ func (sv *openShiftClusterStaticValidator) validateProperties(path string, p *Op
 	if err := sv.validateConsoleProfile(path+".consoleProfile", &p.ConsoleProfile); err != nil {
 		return err
 	}
-	if err := sv.validateServicePrincipalProfile(path+".servicePrincipalProfile", &p.ServicePrincipalProfile); err != nil {
-		return err
+	if isCreate {
```
Comment (jim-minter) [inline pkg/api/v20191231preview/openshiftcluster_validatestatic.go:89]: going to rebase and revert this change, now that servicePrincipalProfile is no longer dynamic.

File: pkg/api/v20191231preview/openshiftcluster_validatestatic.go [hunk 7]
```diff
@@ -86,20 +86,24 @@ func (sv *openShiftClusterStaticValidator) validateProperties(path string, p *Op
...
 	}
 	if err := sv.validateMasterProfile(path+".masterProfile", &p.MasterProfile); err != nil {
 		return err
 	}
-	if len(p.WorkerProfiles) != 1 {
-		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, path+".workerProfiles", "There should be exactly one worker profile.")
-	}
-	if err := sv.validateWorkerProfile(path+".workerProfiles['"+p.WorkerProfiles[0].Name+"']", &p.WorkerProfiles[0], &p.MasterProfile); err != nil {
-		return err
+	if isCreate {
```
Comment (jim-minter) [inline pkg/api/v20191231preview/openshiftcluster_validatestatic.go:100]: these new `isCreate`s don't have enough test coverage. #413 is a follow-up for this.

File: pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go (hunk 3 above)
Comment (jim-minter) [reply pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go:114]: GH needs a "good catch" icon...

File: pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go (hunk 4 above)
Comment (jim-minter) [reply pkg/api/v20191231preview/openshiftcluster_validatestatic_test.go:706]: as a sop to usability.  If the Name field doesn't exist, it reverts to index.

File: pkg/frontend/openshiftcluster_putorpatch_test.go [hunk 8]
```diff
@@ -667,8 +680,9 @@ func TestPutOrPatchOpenShiftCluster(t *testing.T) {
 			asyncOperations := mock_database.NewMockAsyncOperations(controller)
 			openShiftClusters := mock_database.NewMockOpenShiftClusters(controller)
 			subscriptions := mock_database.NewMockSubscriptions(controller)
+			enricher := mock_clusterdata.NewMockOpenShiftClusterEnricher(controller)
```
Comment (jim-minter) [inline pkg/frontend/openshiftcluster_putorpatch_test.go:683]: I think we need a test case where the enricher has a side effect on doc, i.e. using https://godoc.org/github.com/golang/mock/gomock#Call.Do.  https://github.com/Azure/ARO-RP/issues/414.
